openHAB 5.0 release blog post (#521)
openhab/website: blog/2025-07-21-openhab-5-0-release.md

It also included a helper library that simplifies the use of openHAB's API and provides a more Pythonic interface to interact with openHAB's APIs.

It also offers the following improvements over the Jython binding:

- improved error logging
- import native `java.*` or `org.openhab.*` Java packages
- scope module to encapsulate all jsr223 objects
- rule class to simplify rule setup, logging, and more
- registry class which wraps/simplifies access to Items, Things, Channels
- works with native python datatypes like [`datetype`](https://docs.python.org/3/library/datetime.html) instead of [`ZonedDateTime`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/time/ZonedDateTime.html)
- all kinds of triggers, actions and services are supported
- object functions are almost 100% similar to the openHAB Core API
- dependency watcher to reload scripts when modified

The only current limitation is the lack of PIP support for installing external modules, which will be introduced in the next release.
If you don't want to wait, you can install [the beta version via the Marketplace](https://community.openhab.org/t/python-scripting-5-0-0-5-9-9/164947).

> It's also important to note that this binding was not developed as a replacement for HabAPP, which is completely independent, still actively developed, and enjoys a large community.

Reviewer: ```suggestion
It's also important to note that this binding was not developed as a replacement for HABApp, which is completely independent, still actively developed, and enjoys a large community.
```

With openHAB 5.0, we have introduced a drag and drop feature to the sitemap editor, allowing you to easily rearrange elements in the sitemap:

<p align="center"><img style="max-width: 70%;" src="/uploads/2025-07-21-openhab-5-0-release/sitemap-edit-dragdrop.gif"/></p>

Many thanks to Mark Herwege ([@mherwege](https://github.com/mherwege)) for implementing this feature!

### Model Editor Drag & Drop

Similarly to the sitemap editor, the model editor now supports drag and drop for moving semantic tags around.
This not only makes it faster to build your semantic model, it also makes it easier to ensure that your model is valid and follows the best practices by introducing numerous validation checks.

<div align="center" class="row">
  <img style="max-width: 35%; min-width: 250px;" src="/uploads/2025-07-21-openhab-5-0-release/model-dragdrop-1.gif"/>
  <img style="max-width: 35%; min-width: 250px;" src="/uploads/2025-07-21-openhab-5-0-release/model-dragdrop-2.gif"/>
</div>

A big thank you to Mark Herwege ([@mherwege](https://github.com/mherwege)) for working on this game-changing feature!

### Regenerate Rules from Templates

### Copy DSL & YAML Definitions

> 

Author: Thanks, I have just commited this.

  <img style="max-width: 35%; min-width: 250px;" src="/uploads/2025-07-21-openhab-5-0-release/model-dragdrop-2.gif"/>
</div>

A big thank you to Mark Herwege ([@mherwege](https://github.com/mherwege)) for working on this game-changing feature!

### Regenerate Rules from Templates

### Copy DSL & YAML Definition: Connecting UI with Text Configuration

_Jimmy Tanagra ([@jimtng](https://github.com/jimtng)), openHAB Maintainer_

openHAB 5.0 introduces a convenient option for users who wish to convert their setup into text-based configuration.
Definitions of **Items** and **Things** can now be copied in **DSL** or the new **YAML** file format, making it easier to reuse, back up, or migrate content into textual configuration files.

This feature is intended purely as a convenience. Users can continue to rely entirely on the UI configuration if they prefer.

**Where you'll find it:**

- **Things list** and **Items list** – Select multiple entries to reveal a `Copy` button
- **Thing Details** and **Item Details** page – A `Copy File Definition` button is available at the bottom of the page

> - **Developer Tools** – New menu options: `Things File Definition` and `Items File Definition` for one-click access

Reviewer 2: I mean the blue entry in this menu when you are on one entry of the inbox:

<img width="562" height="249" alt="image" src="https://github.com/user-attachments/assets/e827f7b7-783e-4aa5-bcd3-789b0541f469" />

> - **Developer Tools** – New menu options: `Things File Definition` and `Items File Definition` for one-click access

Author: Got it, thanks.

  <img style="max-width: 35%; min-width: 250px;" src="/uploads/2025-07-21-openhab-5-0-release/model-dragdrop-2.gif"/>
</div>

A big thank you to Mark Herwege ([@mherwege](https://github.com/mherwege)) for working on this game-changing feature!

### Regenerate Rules from Templates

### Copy DSL & YAML Definition: Connecting UI with Text Configuration

_Jimmy Tanagra ([@jimtng](https://github.com/jimtng)), openHAB Maintainer_

openHAB 5.0 introduces a convenient option for users who wish to convert their setup into text-based configuration.
Definitions of **Items** and **Things** can now be copied in **DSL** or the new **YAML** file format, making it easier to reuse, back up, or migrate content into textual configuration files.

This feature is intended purely as a convenience. Users can continue to rely entirely on the UI configuration if they prefer.

**Where you'll find it:**

- **Things list** and **Items list** – Select multiple entries to reveal a `Copy` button
- **Thing Details** and **Item Details** page – A `Copy File Definition` button is available at the bottom of the page

> - **Thing Inbox** – A `Copy File Definition` button is available when clicking on an entry in the inbox

Reviewer 3: Thanks @lolodomo for catching this missing one. I'd like to add a little extra hint here, if that's OK
```suggestion
- **Thing Inbox** – Clicking on an inbox entry reveals a `Copy File Definition` button, allowing you to save the discovered Thing in a file-based configuration rather than importing it into the UI-managed database
```

_Cody Cutrer ([@ccutrer](https://github.com/ccutrer)), openHAB Maintainer_

On the subject of Python, the GraalVM work has enabled the Home Assistant MQTT binding to switch from using the Java Jinjava library for emulating Python's Jinja to actually running Python and Jinja.
This allows openHAB to import some of the actual Home Assistant Python code, so template rendering from devices should be near-perfect, without constantly playing a game of catch-up as we find devices that use some feature of Python, Jinja, or Home Assistant that we hadn't previously conceived of.

We've also taken this one step further, and re-implemented parsing of the Home Assistant JSON discovery metadata via Home Assistant's own classes, which fixes myriad strange bugs and incompatibilities from attempting to implement in Java using only Home Assistant documentation as a guide.
We now get the benefit of exactly matching many of those minor idiosyncrasies that aren't documented.
There's still work to do in the portion of the binding that connects incoming data to openHAB channels.
However, it's now sitting on a much stronger base so that any issues there can be more quickly iterated on and fixed.

> 

Reviewer 3: @ccutrer I was just wondering if we could tell people that they can uninstall jinja transform unless they specifically need it outside homeassistant?

Reviewer 4: Wasn't that in 4.3?

Reviewer 3: Ah yes, good memory! I've just re-read the 4.3 blog and it was there! :)

openHAB 5.0 introduces a modular configuration system that allows supported entities (objects such as Things, Items, etc.) to be defined in one or more YAML files.
This new format does not replace our dedicated DSL format, but rather provides an alternative for YAML enthusiasts or users who have never been comfortable with our DSL syntax.
It is also a way to not depend too much on XTend/XText as we don't know how many more years these Eclipse projects will be maintained.
Unlike the DSL configuration files, YAML configuration files are not limited to a specific object type — they can contain any supported entities (Things, Items, Tags, etc.).

Here is an example:

```yaml
version: 1

tags:
  Location_Indoor_Room_HomeCinemaRoom:
    label: Home Cinema Room
    description: The room containing my home cinema system
    synonyms:
      - Home theater room
      - TV room
      - Movie room

things:
  lgwebos:WebOSTV:living-room:
    config:
      host: 192.168.2.119
      key: XXXX-XXXX-XXXX-XXXX
      macAddress: 3c:cd:93:c2:20:e0

items:
  TV_Power:
    type: Switch
    label: TV Power
    icon: television
    autoupdate: false
    channel: lgwebos:WebOSTV:living-room:power
```

> You can consult the [documentation](https://www.openhab.org/docs/configuration/yaml.html).

Reviewer 3: ```suggestion
You can consult the [YAML Configuration Documentation](https://www.openhab.org/docs/configuration/yaml.html).
```

It is now also possible to generate this syntax directly from Main UI if you have Things and Items that were defined either with DSL syntax or created from Main UI, see [Copy DSL & YAML Definition: Connecting UI with Text Configuration](#copy-dsl--yaml-definition-connecting-ui-with-text-configuration)
Support for other entities will be added in the next releases, starting with rules and rule templates.

> Other exciting features based on this are already in the pipe.

Reviewer 3: ```suggestion
Other exciting features based on this are already in the pipeline.
```

  <img style="max-width: 35%; min-width: 165px;" src="/uploads/2025-07-21-openhab-5-0-release/model-dragdrop-2.gif"/>
</div>

A big thank you to Mark Herwege ([@mherwege](https://github.com/mherwege)) for working on this game-changing feature!

### Regenerate Rules from Templates

### Copy DSL & YAML Definition: Connecting UI with Text Configuration

_Jimmy Tanagra ([@jimtng](https://github.com/jimtng)), openHAB Maintainer_

openHAB 5.0 introduces a convenient option for users who wish to convert their setup into text-based configuration.
Definitions of **Items** and **Things** can now be copied in **DSL** or the new **YAML** file format, making it easier to reuse, back up, or migrate content into textual configuration files.

This feature is intended purely as a convenience. Users can continue to rely entirely on the UI configuration if they prefer.

**Where you'll find it:**

- **Things list** and **Items list** – Select multiple entries to reveal a `Copy` button
- **Thing Details** and **Item Details** page – A `Copy File Definition` button is available at the bottom of the page

> - **Thing Inbox** – Clicking on an inbox entry reveals a `Copy File Definition` button, allowing you to save the discovered Thing in a file-based configuration rather than importing it into the UI-managed database

Reviewer 3: Sorry, I've just double checked the exact button label, and it's also available in multiple selection.

```suggestion
- **Things Inbox** – Clicking on an inbox entry — or selecting multiple entries — reveals a Copy Thing File Definition (or Copy) button. This lets you save discovered Things to a file-based configuration instead of importing them into the UI-managed database.
```

Author: No problem!

We have integrated Matter support, added a new Python Scripting add-on, made the semantic model easier to configure and added a new configuration file format.
But that's not all!
As usual for major releases, we have also made improvements under the hood, such as updating the underlying Java version to 21, and making Oracle's GraalVM ready for OSGi to stay up to date with the latest technologies.

As this is a new major release, we have a number of breaking changes that are required for the new features and improvements.
So, please have a look at our [**official release notes**](https://github.com/openhab/openhab-distro/releases/tag/5.0.0) to learn about all breaking changes, new add-ons, enhancements and fixes.

With that being said, we again want to share some statistics that show the progress in numbers and highlight some of the new features that you should not miss.

# Activity

As you'd expect from any new openHAB release, we have had a high activity across all our GitHub repositories.

> We therefore want to share a few statistics about our three main repositories: Core, add-ons, and UIs.

Reviewer 5: Behind every new feature in openHAB 5.0 is the work of a vibrant, global community. Over the past period, activity across our GitHub repositories has been nonstop — a reflection of openHAB’s momentum and shared purpose.

[openHAB Core](https://github.com/openhab/openhab-core) has received xxx pull requests in total, with xxx bug fixes and xxx enhancements resulting in xx,yyy lines of added code.

xx new add-ons, xxx bug fixes and xxx enhancements were contributed in xxx pull requests to the [openHAB add-ons repository](https://github.com/openhab/openhab-addons), which resulted in the addition of xxx,yyy lines of code.

Our [openHAB web UIs](https://github.com/openhab/openhab-webui) have received xxx pull requests, including xxx bug fixes and xxx enhancements resulting in xx,yyy lines of code added.

> It is worth mentioning that [our other repositories](https://github.com/openhab) have also received numerous contributions (e.g. documentation), but listing them all would probably be too much here.
> Many thanks to all our contributors for driving openHAB forward!

Reviewer 5: ```suggestion
Many [other repositories](https://github.com/openhab) — including documentation, openHABian, and build tooling — also receive steady contributions, too numerous to list here.

Huge thanks to everyone who continues to push openHAB forward!
```

Again, it is time to shout out to our top code contributors per repository (having done the most commits) for this new release:

<!-- https://github.com/openhab/openhab-core/graphs/contributors?from=16.12.2024&to=13.7.2025 -->
<!-- https://github.com/openhab/openhab-addons/graphs/contributors?from=16.12.2024&to=13.7.2025 -->
<!-- https://github.com/openhab/openhab-webui/graphs/contributors?from=16.12.2024&to=13.7.2025 -->

- Holger ([@holgerfriedrich](https://github.com/holgerfriedrich)) for openHAB Core
- Jim ([@jimtng](https://github.com/jimtng)) for openHAB Core
- Jacob ([@jlaur](https://github.com/jlaur)) for openHAB add-ons
- Andrew ([@andrewfg](https://github.com/andrewfg)) for openHAB add-ons
- Florian ([@florian-h05](https://github.com/florian-h05)) for the openHAB Main UI
- Jim ([@jimtng](https://github.com/jimtng)) for the openHAB Main UI

Many of these contributors also serve as maintainers — reviewing pull requests, discussing architectural changes, and supporting newer contributors.

> At the time of this release, we have **39 active maintainers** across all repositories.

Author: Hmm, that number is much more difficult to get across all repos ... we can check core, addons, webui, docs and should "catch" most contributors that way.

> At the time of this release, we have **39 active maintainers** across all repositories.

Author: I have created https://github.com/florian-h05/github-stats for that task.

Our numbers are: 139 unique contributors, 2465 commits.

> At the time of this release, we have **39 active maintainers** across all repositories.

Author: BTW just refactored it and extended it to also give the numbers of enhancements, but fixes etc and the top contributors.

> Many of our top committers are also maintainers, which means besides contributing code, they also review pull requests and participate in discussions on GitHub issues, which helps maintain our high level of quality across bindings, UIs and Core.
> At present, we have 39 maintainers across all our repositories.
> Huge thanks to all our contributors and all my maintainer colleagues for their work on our next major release!

Reviewer 5: ```suggestion
Many of these contributors also serve as maintainers — reviewing pull requests, discussing architectural changes, and supporting newer contributors.
At the time of this release, we have **39 active maintainers** across all repositories.

Whether fixing bugs, refining UI elements, updating documentation, or building entirely new integrations — every contribution matters.
```

It offers many improvements over the Jython binding, including:

- more convenient access to Items, Things & channels through the helper library
- a simplified way of creating and managing rules through the helper library
- support for native Python datatypes like [`datetype`](https://docs.python.org/3/library/datetime.html) instead of [`ZonedDateTime`](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/time/ZonedDateTime.html)
- a dependency watcher that reloads scripts when dependencies change

The only current limitation is the lack of PIP support for installing external modules, which will be introduced in the next release.
If you don't want to wait, you can install [the beta version via the Marketplace](https://community.openhab.org/t/python-scripting-5-0-0-5-9-9/164947).

It's also important to note that this binding was not developed as a replacement for HABApp, which is completely independent, still actively developed, and enjoys a large community.
Your choice among these two options depends entirely on your personal preference.

## openHAB iOS App

_Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), openHAB Maintainer_

> The openHAB iOS app will soon receive a major update including major enhancements, both in its core code as well a new features.

Reviewer 6: ```suggestion
The openHAB iOS app will soon receive a major update, including significant enhancements to its core code as well as introducing new features.
```

Major smart home companies like Apple, Google, Amazon, Samsung SmartThings, Signify (Philips Hue) and IKEA have invested significant resources jointly developing the Matter standard and have included Matter in nearly every device they ship.
With updates and additions to the Matter specification happening twice a year, Matter's adoption in the market is growing rapidly as it constantly adapts to support new devices, new use-cases and improves with feedback from users and vendors alike.

### openHAB Matter Client

<style>
.matter-screenshot {
  float: right;
  max-width: 25% !important;
  margin-left: 1em;

  @media (max-width: 768px) {
    min-width: 330px;
    float: none;
    margin-left: 0;
  }

  @media (max-width: 1024px) {
    max-width: 40% !important;
  }
}
</style>

<img
  src="/uploads/2025-07-21-openhab-5-0-release/matter-thing.png"
  class="matter-screenshot"
  alt="Matter thing"
/>

> openHAB 5.0 now features one of the most complete Matter 1.4.1 client implementations on the market — the [Matter binding](/addons/bindings/matter), which can communicate with more than 30 types of devices, from lights, locks and thermostats, to power meters, air quality sensors and robotic vacuum cleaners.Matter Devices can be Wi-Fi or ethernet based, or they can be based on [Thread](https://www.threadgroup.org/), a low-power radio technology and successor to the ZigBee protocol which is baked into many recent consumer products like Apple TVs, Google Home Hubs, Amazon Echos as well as many Wi-Fi access points and routers.

Reviewer 6: ```suggestion
openHAB 5.0 now features one of the most complete Matter 1.4.1 client implementations on the market — the [Matter binding](/addons/bindings/matter), which can communicate with more than 30 types of devices, from lights, locks and thermostats, to power meters, air quality sensors and robotic vacuum cleaners.
Matter Devices can be Wi-Fi or ethernet based, or they can be based on [Thread](https://www.threadgroup.org/), a low-power radio technology and successor to the ZigBee protocol which is baked into many recent consumer products like Apple TVs, Google Home Hubs, Amazon Echos as well as many Wi-Fi access points and routers.
```

- **Widget Config:** The widget config popup has been redesigned to support live updates, dirty checking and be movable, so you can watch the changes in your page while editing the widget config.
- **Specific Browser Titles:** The browser title now reflects the current page, making it easier to identify the contents of the openHAB tabs in your browser.
- **Log Viewer Enhancements:** The log viewer has been enhanced with a history buffer, allowing you to scroll back through previous log entries, and a log details popup that provides more information about each log entry.
  Additionally, a text mode has been introduced as alternative to the default table view, having the look and feel of the well-known [frontail](https://github.com/mthenw/frontail) log viewer.
- **Thing Inbox:** Add search, approve multiple Things at once and allow to change Thing ID & label when approving Things.

## Blockly Enhancements

- New Rule-Enable Block to enable or disable rules.
- New dictionary block to update dictionary entry values
- Persistence Block: Riemann sum support added
- Boolean-typed variable added
- Get Item Previous State added

## Sitemap & Basic UI Enhancements

> - New `interpolation` parameter for charts to change how the line between two data points is drawn: Either `linear` or `step`.

Reviewer 2: Please remove this line, it looks like this sitemap feature is not yet implemented in any sitemap UI!

Author: Done.

</div>

Similarly to the sitemap editor, the model editor now supports drag and drop for moving semantic tags around.
This not only makes it faster to build your semantic model, it also makes it easier to ensure that your model is valid and follows the best practices by introducing numerous validation checks.

A big thank you to Mark Herwege ([@mherwege](https://github.com/mherwege)) for working on this game-changing feature!
<div style="clear: both;"></div>

### Regenerate Rules from Templates

_Ravi Nadahar ([@Nadahar](https://)), openHAB Contributor_

Previously, rules created from rule templates were static and could not be updated if the template was changed.
It was necessary to delete and recreate the rules from the template to apply any updates or changes to the parameters used in the template.
openHAB 5.0 introduces a new feature that allows you to regenerate rules from templates, making it easier to keep your rules up to date with the latest changes in the templates.
This also allows you to change the parameters used in the template without having to delete and recreate the rules.

> You can find more details in [the documentation](/docs/tutorial/rules_basic.html#updates-to-templates).

Reviewer 7: ```suggestion
You can find more details in [the documentation](/docs/tutorial/rules_basic.html#update-a-rule).
```
Ideally, the link should be this instead. Sorry for not doing it right the first time, but it's not "wrong" the way it is either.

- TODO

Many of our top committers are also maintainers, which means besides contributing code, they also review pull requests and participate in discussions on GitHub issues, which helps maintain our high level of quality across bindings, UIs and Core.
At present, we have xx maintainers across all our repositories.
Huge thanks to all our contributors and all my maintainer colleagues for their work on our next major release!

# Highlights

As you might have already noticed in the activity statistics, there was much development on openHAB - too much to list everything.
Please refer to the [**release notes**](https://github.com/openhab/openhab-distro/releases/tag/5.0.0) to check out what's new and noteworthy (e.g. breaking changes).

In the following sections, our maintainers and contributors are introducing some new features you don't want to miss!

## Matter Integration

_Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), openHAB Maintainer_

## Python Scripting

_Holger Hees ([@HolgerHees](https://github.com/HolgerHees)), openHAB Maintainer_

## Z-Wave JS Binding

> 

Reviewer 2: Yes, that is in my opinion a very good balance. Perfect.

> 

Reviewer 2: I would just expect something a little bigger for the Matter binding.
